File: Data Structures & Algorithms/two-integer-sum-ii/submission-1.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        vector<int> ret(2);
        int l = 0 , r = numbers.size() - 1;
        while (l < r){
            if (target > numbers[l] + numbers[r])   l++;
            else if (target < numbers[l] + numbers[r])   r--;
            else{
                if (l == r) return ret;
                ret[0] = l + 1;
                ret[1] = r + 1;
                return ret;
            }
        }
    }
};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.cs
public class Solution {
    public List<List<int>> CombinationSum2(int[] candidates, int target) {
        Array.Sort(candidates);
        foreach(var candidate in candidates) Console.WriteLine(candidate);
        List<List<int>> ret = new List<List<int>>();
        
        dfs(candidates, ref ret, new List<int>(), target, 0);
        return ret;
    }

    void dfs(int[] candidates, ref List<List<int>> ret, List<int> visited, int target, int idx){
        if (target == 0)  {
            ret.Add(new List<int>(visited));
            return;
        }if (target < 0) {            
            return;
        }
        if (idx >= candidates.Length) {          
            return;
        }
        
        int next_idx = idx;

        while(next_idx < candidates.Length && candidates[idx] == candidates[next_idx]) next_idx++;
        //skip current element
        dfs(candidates, ref ret, visited, target, next_idx);

        //take the current element
        visited.Add(candidates[idx]);
        dfs(candidates, ref ret, visited, target - candidates[idx], idx + 1);
        visited.RemoveAt(visited.Count - 1);
    }
}
